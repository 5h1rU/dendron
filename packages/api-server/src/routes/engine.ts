import {
  DEngineV2,
  EngineQueryRequest,
  EngineWriteRequest,
} from "@dendronhq/common-all";
import { DendronEngine } from "@dendronhq/engine-server";
import { Request, Response, Router } from "express";
import { MemoryStore } from "../store/memoryStore";

const router = Router();

router.get("/health", async (_req: Request, res: Response) => {
  return res.json({ ok: 1 });
});

router.post("/query", async (req: Request, res: Response) => {
  const { ws, queryString, mode } = req.body as EngineQueryRequest;
  const engine = await MemoryStore.instance().get<DendronEngine>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  const { error, data } = await engine.query(queryString, mode);
  const cleanData = data.map((ent) => ent.toRawProps());
  res.json({ error, data: cleanData });
});

router.post("/write", async (req: Request, res: Response) => {
  const { ws, node, opts } = req.body as EngineWriteRequest;
  const engine = await MemoryStore.instance().get<DEngineV2>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  try {
    await engine.writeNote(node, opts);
    res.json({ error: null, data: null });
  } catch (err) {
    res.json({ error: JSON.stringify(err), data: null });
  }
});

export { router as engineRouter };
