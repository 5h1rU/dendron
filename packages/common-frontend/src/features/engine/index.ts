export * as engineReducer from "./slice";
export {
  setError,
  setNotes,
  EngineSliceUtils,
  initNotes,
  InitNoteOpts,
  EngineState,
} from "./slice";
export * as engineStore from "./store";
export {
  RootState as engineState,
  AppDispatch as engineDispatch,
} from "./store";
