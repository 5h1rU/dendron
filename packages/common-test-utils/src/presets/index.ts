export * from "./initialize";
export * from "./pods-core";
export * from "./engine-server";
export * from "./plugin-core";
export * from "./utils";
export * from "./notes";
