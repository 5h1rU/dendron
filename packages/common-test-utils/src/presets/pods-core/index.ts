import SNAPSHOT from "./snapshot";
import JSON from "./json";

export const PODS_CORE = {
  SNAPSHOT,
  JSON,
};
