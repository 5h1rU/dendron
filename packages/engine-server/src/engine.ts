import {
  DEngine,
  DEngineStore,
  DNodeData,
  DNodeDict,
  DNodeRawProps,
  EngineQueryResp,
  IDNode,
  INoteOpts,
  NodeBuilder,
  NodeWriteOpts,
  Note,
  NoteDict,
  QueryMode,
  QueryOpts,
  Schema,
  SchemaDict,
  SchemaRawProps,
  Scope,
  assert,
  assertExists,
  createLogger,
  getStage,
  makeResponse
} from "@dendronhq/common-all";

import { BodyParser } from "./drivers/raw/BodyParser";
import FileStorage from "./drivers/file/store";
import Fuse from "fuse.js";
import _ from "lodash";
import fs from "fs-extra";

let PROTO_ENGINE: ProtoEngine;
const logger = createLogger("DEngine");

function isAllQuery(qs: string): boolean {
  return qs === "**/*";
}

export interface FuseOptions {
  exactMatch?: boolean;
  preset: QueryMode;
  // isCaseSensitive?: boolean
  // distance?: number
  // findAllMatches?: boolean
  // includeMatches?: boolean
  // includeScore?: boolean
  // location?: number
  // minMatchCharLength?: number
  // shouldSort?: boolean
  // threshold?: number
  // useExtendedSearch?: boolean
}

function createFuse<T>(initList: T[], opts: FuseOptions) {
  const options = {
    shouldSort: true,
    threshold: opts.exactMatch ? 0.0 : 0.6,
    location: 0,
    distance: 100,
    maxPatternLength: 32,
    minMatchCharLength: 1,
    keys: ["title", "path"]
  };
  if (opts.preset === "schema") {
    options.keys = ["title", "id"];
  }
  // initList = _.map(initList, (n) => ({ ...n, treePath: n.path }));
  const fuse = new Fuse(initList, options);
  return fuse;
}

type ProtoEngineOpts = {
  root?: string;
  cacheDir?: string;
  forceNew?: boolean;
};

type ProtoEngineProps = Required<ProtoEngineOpts>;

export class ProtoEngine implements DEngine {
  public fuse: Fuse<Note, any>;
  public schemaFuse: Fuse<Schema, any>;

  public notes: NoteDict;
  public schemas: SchemaDict;

  public fullNodes: Set<string>;

  public opts: ProtoEngineProps;

  public queries: Set<string>;

  public store: DEngineStore;

  static getEngine(opts?: ProtoEngineOpts): DEngine {
    if (!PROTO_ENGINE || opts?.forceNew) {
      const stage = getStage();
      opts = _.defaults(opts || {}, {
        // TODO: remove hardcoded value
        root: "/Users/kevinlin/Dropbox/Apps/Noah/notesv2"
      });
      if (_.isUndefined(opts.root)) {
        throw Error(`root must be defined`);
      }
      // TODO
      PROTO_ENGINE = new ProtoEngine(
        new FileStorage(opts as { root: string }),
        opts
      );
      logger.info({ ctx: "getEngine:exit", opts, stage });
      return PROTO_ENGINE;
    }
    return PROTO_ENGINE;
  }

  constructor(store: DEngineStore, opts: ProtoEngineOpts) {
    // eslint-disable-next-line spaced-comment
    //this.nodes = store.fetchInitial();
    this.notes = {};
    this.store = store;
    this.fuse = createFuse<Note>([], { exactMatch: false, preset: "note" });
    this.fullNodes = new Set();
    this.queries = new Set();

    // setup schemas
    this.schemas = {};
    this.schemaFuse = createFuse<Schema>([], {
      exactMatch: false,
      preset: "schema"
    });

    this.opts = _.defaults(opts, {
      cacheDir: "/tmp/dendronCache",
      root: "/Users/kevinlin/Dropbox/Apps/Dendron",
      forceNew: false
    });
    [opts.cacheDir, opts.root].forEach(fpath => {
      if (!fs.existsSync(fpath as string)) {
        throw Error(`${fpath} doesn't exist`);
      }
    });
  }

  deleteFromNodes(id: string) {
    delete this.notes[id];
    this.fullNodes.delete(id);
  }

  /**
   * Updates local cache
   * - if node is new, add it to cache
   * - if not new, update existing node properties
   * - extra:
   *   - if full node, add id
   * @param nodes
   * @param opts
   */
  refreshNodes(nodes: IDNode[], opts?: QueryOpts) {
    if (_.isEmpty(nodes)) {
      return;
    }
    const mode = nodes[0].type;
    if (mode === "schema") {
      const schemas = nodes as Schema[];
      schemas.forEach((node: Schema) => {
        this.schemas[node.id] = node;
      });
      this.updateLocalCollection(_.values(this.schemas), "schema");
      return;
    } else {
      nodes.forEach((node: IDNode) => {
        const { id } = node;

        if (!_.has(this.notes, id)) {
          // add if not exist
          // TODO: nodes has both raw and full nodes
          // @ts-ignore
          this.notes[id] = node;
        } else {
          // exists, merge it
          logger.debug({
            ctx: "refreshNodes:existingNode",
            node: node.toRawProps()
          });
          _.merge(this.notes[id], node);
        }
        // if a full node, make sure to add
        if (opts?.fullNode) {
          this.fullNodes.add(id);
        }
      });
      // FIXME: debug
      const newNodes = _.map(nodes, n => ({ title: n.title, id: n.id }));
      const allNodes = _.map(this.notes, n => ({
        title: n.title,
        id: n.id
      }));
      logger.debug({
        ctx: "refreshNodes",
        newNodes,
        allNodes
      });
      this.updateLocalCollection(_.values(this.notes), "note");
      return;
    }
  }

  /**
   * Turn parent|children ids to full nodes
   */
  async resolveIds(
    node: DNodeRawProps<DNodeData>,
    dict: DNodeDict
  ): Promise<Note> {
    const nodeResolved: INoteOpts = dict[node.id] as INoteOpts;
    assert(
      !_.isUndefined(nodeResolved),
      `node ${JSON.stringify(node)} not in dict\nkeys: ${_.keys(dict)}`
    );
    nodeResolved.body = node.body;
    return new Note({ ...nodeResolved });
  }

  updateLocalCollection(collection: IDNode[], mode: QueryMode) {
    if (mode === "schema") {
      return this.schemaFuse.setCollection(collection as Schema[]);
    } else {
      return this.fuse.setCollection(collection as Note[]);
    }
  }

  async delete(id: string): Promise<void> {
    await this.store.delete(id);
    this.deleteFromNodes(id);
    return;
  }

  /**
   * Defaults
   * @param id
   * @param opts
   *  - fullNode: {default:true}
   */
  async get(_scope: Scope, id: string, mode: QueryMode, opts?: QueryOpts) {
    opts = _.defaults(opts || {}, { fullNode: true, createIfNew: true });
    let nodeDict;
    logger.debug({ ctx: "get", id, opts });

    if (mode === "schema") {
      nodeDict = this.schemas;
      const node = nodeDict[id];
      return { data: node };
    }
    //FIXME: check if exist
    // we are assuming all nodes are fetched
    const node = this.notes[id];

    // a full node has a body and is fully resolved
    if (opts?.fullNode && !this.fullNodes.has(id)) {
      logger.debug({ ctx: "get:fetchFromStore:pre", id });
      const fnResp = await this.store.get(_scope, id, {
        ...opts,
        webClient: true
      });
      logger.debug({ ctx: "get:fetchFromStore:post", id, opts, fnResp });
      const fullNode = await this.resolveIds(fnResp.data, this.notes);
      logger.debug({ ctx: "get:resolve:post", fnResp });
      // TODO:
      this.refreshNodes([fullNode], opts);
      return { data: fullNode };
    }
    logger.debug({ ctx: "get:exit", node });
    return { data: node };
  }

  async query(
    scope: Scope,
    queryString: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) {
    opts = _.defaults(opts || {}, {
      fullNode: false,
      createIfNew: true,
      initialQuery: false
    });
    let data: EngineQueryResp;

    // handle all query case
    if (isAllQuery(queryString)) {
      logger.debug({ ctx: "query:queryAll:pre", mode });
      data = await this.store.query(scope, "**/*", mode, opts);
      if (opts.initialQuery) {
        this.refreshNodes(data.data);
      }
      // @ts-ignore
      return data as EngineQueryResp;
    }

    // handle schema query
    if (mode === "schema") {
      const results = this.schemaFuse.search(queryString);
      let items: Schema[];
      const nodes = this.schemas;
      if (opts.queryOne) {
        items = [results[0]?.item];
      } else {
        items = _.map(results, resp => resp.item);
      }
      logger.debug({ ctx: "query:exit:schema", items });
      return makeResponse<EngineQueryResp>({
        data: _.map(items, item => nodes[item.id]),
        error: null
      });
    } else {
      // handle note query
      const results = this.fuse.search(queryString);
      let items: IDNode[];
      if (opts.queryOne) {
        items = [results[0]?.item];
      } else {
        items = _.map(results, resp => resp.item);
      }

      if (opts.queryOne && items[0]?.path !== queryString && opts.createIfNew) {
        logger.debug({
          ctx: "query:write:pre",
          queryString,
          item: items[0]
        });
        const nodeBlank = new Note({ fname: queryString });
        await this.write(scope, nodeBlank, { newNode: true });
        const { data: nodeFull } = await this.get(scope, nodeBlank.id, mode);
        this.refreshNodes([nodeFull], { fullNode: true });
        return makeResponse<EngineQueryResp>({
          data: [nodeFull],
          error: null
        });
      }

      if (opts.fullNode) {
        const fetchedFullNodes = await Promise.all(
          _.map<IDNode, Promise<IDNode | null>>(items, async ent => {
            if (!this.fullNodes.has(ent.id)) {
              logger.debug({
                ctx: "query:fuse.search:post",
                status: "fetch full node from store"
              });
              // FIXME: ratelimit
              const fn = await this.get(scope, ent.id, mode);
              return fn.data;
            }
            logger.debug({
              ctx: "query:fuse.search:post",
              status: "fetch full node from cache"
            });
            return null;
          })
        );
        this.refreshNodes(
          _.filter(fetchedFullNodes, ent => !_.isNull(ent)) as IDNode[],
          { fullNode: true }
        );
        logger.debug({
          ctx: "query:fetchedFullNodes:exit",
          fetchedFullNodes
        });
      }
      logger.debug({ ctx: "query:exit:note", items });
      return makeResponse<EngineQueryResp>({
        data: _.map(items, item => this.notes[item.id]),
        error: null
      });
    }
  }

  async write(scope: Scope, node: IDNode, opts?: NodeWriteOpts): Promise<void> {
    opts = _.defaults(opts, { newNode: false, body: "" });
    if (node.type === "schema") {
      assertExists(opts.body, "body must exist");
      // convert body
      const props: SchemaRawProps[] = new BodyParser().parseSchema(
        opts.body as string,
        {
          node,
          fname: node.fname
        }
      );
      const schemas = new NodeBuilder().buildSchemaFromProps(props);
      const schemaNew = _.find(schemas, { id: node.id });
      if (_.isUndefined(schemaNew)) {
        throw Error(`no schema found for ${node} on write`);
      }
      node = schemaNew;
      // reset body
      node.body = "";
      await this.store.write(scope, node);
      if (opts.newNode) {
        const parentPath = "root";
        const parentNode = _.find(this.schemas, n => n.title === parentPath);
        if (!parentNode) {
          throw Error("no parent found");
        }
        (parentNode as Schema).addChild(node as Schema);
      }
    } else {
      const note = node as Note;
      await this.store.write(scope, note);
      if (opts.newNode) {
        let parentPath = note.fname
          .split(".")
          .slice(0, -1)
          .join(".");
        if (_.isEmpty(parentPath)) {
          parentPath = "root";
        }
        const parentNode = _.find(this.notes, n => n.path === parentPath);
        if (!parentNode) {
          throw Error("no parent found");
        }
        parentNode.addChild(note);
      }
    }
    // TODO
    return this.refreshNodes([node], { fullNode: opts.newNode });
  }
}

export function engine() {
  return ProtoEngine.getEngine();
}
