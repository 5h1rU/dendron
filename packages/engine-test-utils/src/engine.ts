import { DEngineClientV2, DVault, WorkspaceOpts } from "@dendronhq/common-all";
import { launch } from "@dendronhq/api-server";
import { tmpDir } from "@dendronhq/common-server";
import {
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
} from "@dendronhq/common-test-utils";
import {
  createEngine as engineServerCreateEngine,
  DendronEngineClient,
  WorkspaceService,
} from "@dendronhq/engine-server";
import _ from "lodash";

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<DEngineClientV2>;

export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return engineServerCreateEngine(opts);
}

export { DEngineClientV2, DVault, WorkspaceOpts };

export async function createEngineFromServer(opts: WorkspaceOpts) {
  const port = await launch({});
  const engine: DEngineClientV2 = DendronEngineClient.create({
    port,
    vaultsv4: opts.vaults,
    ws: opts.wsRoot,
    vaults: [],
  });
  await engine.init();
  return engine;
}

async function setupWS(opts: { vaults: DVault[] }) {
  const wsRoot = tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  const vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault });
      return vault;
    })
  );
  return { wsRoot, vaults };
}

// @ts-ignore
export async function runEngineTestV5<TExtra = any>(
  func: RunEngineTestFunctionV4,
  opts: {
    preSetupHook?: SetupHookFunction;
    //postSetupHook?: PostSetupHookFunction;
    createEngine?: AsyncCreateEngineFunction;
    extra?: TExtra;
    expect: any;
    vaults?: DVault[];
    setupOnly?: boolean;
  }
): Promise<any> {
  const { preSetupHook, extra, vaults, createEngine } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    vaults: [{ fsPath: "vault1" }, { fsPath: "vault2" }],
  });
  const { wsRoot } = await setupWS({ vaults });
  await preSetupHook({ wsRoot, vaults });
  const engine = await createEngine({ wsRoot, vaults });
  const initResp = await engine.init();
  const testOpts = { wsRoot, vaults, engine, initResp, extra };
  if (opts.setupOnly) {
    return testOpts;
  }
  const results = (await func(testOpts)) || [];
  await runJestHarnessV2(results, expect);
  return { opts: testOpts, resp: undefined };
}
