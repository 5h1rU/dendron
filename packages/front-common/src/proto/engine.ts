import {
  DEngine,
  DEngineStore,
  DNodeDict,
  IDNode,
  NodeGetResp,
  NodeQueryResp,
  QueryOpts,
  Scope,
} from "../common/types";
import { DropboxStorage, Note, makeResponse } from "@dendron/common-all";

import Fuse from "fuse.js";
import { Logger } from "@aws-amplify/core";
import _ from "lodash";

const logger = new Logger("DEngine");

let PROTO_ENGINE: ProtoEngine;
function createMockData() {
  const secondChildNote = new Note({
    id: "manifesto",
    title: "manifesto",
    desc: "first child desc",
    type: "note",
    schemaId: "-1",
    // body: "Dendron Manifesto",
  });
  const firstChildNote = new Note({
    id: "dendron",
    title: "dendron",
    desc: "first child desc",
    type: "note",
    schemaId: "-1",
    // body: "Dendron Project",
  });
  const rootNote = new Note({
    id: "root",
    title: "root",
    desc: "root",
    type: "note",
    schemaId: "-1",
    // body: "The root node",
  });
  rootNote.addChild(firstChildNote);
  firstChildNote.addChild(secondChildNote);

  const initialNodes: DNodeDict = {
    [rootNote.id]: rootNote,
    [firstChildNote.id]: firstChildNote,
    [secondChildNote.id]: secondChildNote,
  };
  return initialNodes;
}
const INITIAL_DATA = createMockData();

export interface FuseOptions {
  exactMatch?: boolean;
  // isCaseSensitive?: boolean
  // distance?: number
  // findAllMatches?: boolean
  // includeMatches?: boolean
  // includeScore?: boolean
  // location?: number
  // minMatchCharLength?: number
  // shouldSort?: boolean
  // threshold?: number
  // useExtendedSearch?: boolean
}

function createFuse(initList: IDNode[], opts: FuseOptions) {
  const options = {
    shouldSort: true,
    threshold: opts.exactMatch ? 0.0 : 0.6,
    location: 0,
    distance: 100,
    maxPatternLength: 32,
    minMatchCharLength: 1,
    keys: ["title", "path"],
  };
  // initList = _.map(initList, (n) => ({ ...n, treePath: n.path }));
  // console.log({ initList, bond: true });
  const fuse = new Fuse(initList, options);
  return fuse;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// @ts-ignore
export class MockDataStore implements DEngineStore {
  public data: DNodeDict;
  constructor() {
    this.data = INITIAL_DATA;
  }

  fetchInitial() {
    return this.data;
  }

  async get(_scope: Scope, id: string): Promise<NodeGetResp> {
    // TODO
    return new Promise((resolve) => {
      setTimeout(() => {
        const note = _.get(
          INITIAL_DATA,
          id,
          new Note({
            id,
            title: "sample",
            desc: "sample",
            type: "note",
            schemaId: "-1",
          })
        );
        note.body = `content for ${id}`;
        resolve({ data: note });
      }, 1200);
    });
  }
}

// @ts-ignore - TODO: implement interface
export class ProtoEngine implements DEngine {
  public fuse: Fuse<IDNode, any>;
  public nodes: DNodeDict;
  public fullNodes: Set<string>;
  public queries: Set<string>;
  public store: DEngineStore;

  static getEngine(): DEngine {
    if (!PROTO_ENGINE) {
      // PROTO_ENGINE = new ProtoEngine(new MockDataStore());
      PROTO_ENGINE = new ProtoEngine(new DropboxStorage());
      return PROTO_ENGINE;
    }
    return PROTO_ENGINE;
  }

  constructor(store: DEngineStore) {
    //this.nodes = store.fetchInitial();
    this.nodes = {};
    this.store = store;
    const fuseList = _.values(this.nodes);
    this.fuse = createFuse(fuseList, { exactMatch: false });
    this.fullNodes = new Set();
    this.queries = new Set();
  }

  _nodeInCache(node: IDNode, opts?: QueryOpts) {
    const hasStub = _.has(this.nodes, node.id);
    const fufillsFull = opts?.fullNode ? true : this.fullNodes.has(node.id);
    return hasStub && fufillsFull;
  }

  // FIXME: query doesn't check for full nodes
  _queryInCache(qs: string) {
    const hasQuery = _.has(this.queries, qs);
    return hasQuery;
  }

  /**
   * Updates local cache
   * @param nodes
   * @param opts
   */
  refreshNodes(nodes: IDNode[], opts?: QueryOpts) {
    nodes.forEach((node: IDNode) => {
      const { id } = node;
      // add if not exist
      if (!_.has(this.nodes, id)) {
        this.nodes[id] = node;
      } else {
        _.merge(this.nodes[id], node);
        if (opts?.fullNode) {
          this.fullNodes.add(id);
        }
      }
    });
    this.updateLocalCollection(_.values(this.nodes));
  }

  updateLocalCollection(collection: IDNode[]) {
    this.fuse.setCollection(collection);
  }

  async get(_scope: Scope, id: string, opts?: QueryOpts) {
    //FIXME: check if exist
    const node = this.nodes[id];
    opts = _.defaults(opts || {}, { fullNode: true });
    if (opts?.fullNode && !this.fullNodes.has(id)) {
      const fnResp = await this.store.get(_scope, id, {
        ...opts,
        webClient: true,
      });
      logger.debug({ ctx: "get:store.get:post", id, opts, fnResp });
      this.refreshNodes([fnResp.data], opts);
      return fnResp;
    } else {
      return { data: node };
    }
  }

  async query(scope: Scope, queryString: string, opts?: QueryOpts) {
    console.log({ scope, queryString, opts });
    opts = _.defaults(opts || {}, {
      fullNode: false,
    });
    // TODO: hack
    if (queryString === "**/*") {
      const data = await this.store.query(scope, "**/*", opts);
      this.refreshNodes(data.data);
      return data;
    }
    // FIXME: assuem we have everything
    // if (!this._queryInCache(queryString)) {
    //   await this.store.query(scope, queryString, opts);
    // }
    // TODO: fetch remote
    const results = this.fuse.search(queryString);
    logger.debug({ ctx: "query:fuse.search:post", results, queryString });
    let items: IDNode[];
    if (opts.queryOne) {
      items = [results[0].item];
    } else {
      items = _.map(results, (resp) => resp.item);
    }

    if (opts.fullNode) {
      const fetchedFullNodes = await Promise.all(
        _.map<IDNode, Promise<IDNode | null>>(items, async (ent) => {
          if (!this.fullNodes.has(ent.id)) {
            logger.debug({
              ctx: "query:fuse.search:post",
              status: "fetch full node from store",
            });
            // FIXME: ratelimit
            const fn = await this.get(scope, ent.id);
            return fn.data;
          } else {
            logger.debug({
              ctx: "query:fuse.search:post",
              status: "fetch full node from cache",
            });
            return null;
          }
        })
      );
      this.refreshNodes(
        _.filter(fetchedFullNodes, (ent) => !_.isNull(ent)) as IDNode[],
        { fullNode: true }
      );
      logger.debug({ ctx: "query:fetchedFullNodes:exit", fetchedFullNodes });
    }
    logger.debug({ ctx: "query:exit", items });
    return makeResponse<NodeQueryResp>({
      data: _.map(items, (item) => this.nodes[item.id]),
      error: null,
    });
  }
}

export function engine() {
  return ProtoEngine.getEngine();
}
