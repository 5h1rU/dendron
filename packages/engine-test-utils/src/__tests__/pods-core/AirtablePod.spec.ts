import {
  axios,
  AxiosError,
  DendronError,
  DEngineClient,
  ErrorUtils,
  StatusCodes,
  Time,
  WorkspaceOpts,
} from "@dendronhq/common-all";
import { tmpDir } from "@dendronhq/common-server";
import { AirtableExportPod } from "@dendronhq/pods-core";
import fs from "fs-extra";
import path from "path";
import sinon from "sinon";
import { runEngineTestV5 } from "../../engine";
import { ENGINE_HOOKS } from "../../presets";

const runExport = (opts: WorkspaceOpts & { engine: DEngineClient }) => {
  const pod = new AirtableExportPod();
  const srcHierarchy = "foo";
  return pod.execute({
    ...opts,
    config: {
      dest: "TODO",
      apiKey: "fakeKey",
      baseId: "fakeBase",
      tableName: "fakeTable",
      srcFieldMapping: {
        Title: "title",
        "Updated On": "updated",
        Notes: "body",
      },
      srcHierarchy,
    },
  });
};
const createAxiosError = ({
  response,
}: {
  response?: Partial<AxiosError["response"]>;
}) => {
  let err: AxiosError = {
    isAxiosError: true,
    config: {},
    // @ts-ignore
    response,
    message: "error",
    name: "error",
    toJSON: () => {
      return {};
    },
  };
  return err;
};

describe("GIVEN airtable export", () => {
  describe("WHEN error from post", () => {
    afterEach(() => {
      sinon.restore();
    });
    test("THEN throw error", async () => {
      await runEngineTestV5(
        async (opts) => {
          try {
            sinon
              .stub(axios, "post")
              .rejects(new DendronError({ message: "foo" }));
            await runExport(opts);
          } catch (err) {
            expect(ErrorUtils.isDendronError(err)).toBeTruthy();
          }
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });

  describe.skip("WHEN bad field mapping", () => {
    beforeEach(() => {
      sinon.reset();
    });
    afterEach(() => {
      sinon.restore();
    });

    test("THEN throw detailed error", async () => {
      await runEngineTestV5(
        async (opts) => {
          try {
            const err = createAxiosError({
              response: {
                //
                data: {
                  type: "UNKNOWN_FIELD_NAME",
                  message: 'Unknown field name: "Title"',
                },
                status: StatusCodes.UNPROCESSABLE_ENTITY,
              },
            });
            sinon.stub(axios, "post").rejects(err);
            await runExport(opts);
          } catch (err) {
            expect(ErrorUtils.isDendronError(err)).toBeTruthy();
            expect((err as DendronError).message).toEqual(
              'Unknown field name: "Title"'
            );
          }
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });
});

describe("checkpointing ", () => {
  let exportDest: string;
  let checkpoint: string;
  let timestamp: number;

  beforeEach(() => {
    exportDest = tmpDir().name;
    const basePath = path.dirname(exportDest);
    checkpoint = path.join(
      basePath,
      "pods",
      "dendron.airtable",
      "airtable-pod.lastupdate",
      "checkpoint.txt"
    );
    fs.ensureDirSync(path.dirname(checkpoint));
  });

  test("checkpointing file with greater timestamp", async () => {
    await runEngineTestV5(
      async ({ engine, vaults, wsRoot }) => {
        const pods = new AirtableExportPod();
        const srcHierarchy = "foo";
        pods.processNote = jest.fn();
        timestamp = Time.now().toMillis() + 50000000;
        fs.writeFileSync(checkpoint, timestamp.toString(), {
          encoding: "utf8",
        });

        const resp = async () => {
          return await pods.execute({
            engine,
            vaults,
            wsRoot,
            config: {
              dest: exportDest,
              apiKey: "apikey",
              baseId: "baseId",
              tableName: "Dendron",
              srcFieldMapping: {
                Title: "title",
                "Updated On": "updated",
                Notes: "body",
              },
              srcHierarchy,
            },
          });
        };
        expect.assertions(1);
        return expect(resp()).rejects.toEqual(
          Error(
            "No new Records to sync in selected hierarchy. Create new file and then try"
          )
        );
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
      }
    );
  });

  test("checkpointing file with lower timestamp", async () => {
    await runEngineTestV5(
      async ({ engine, vaults, wsRoot }) => {
        const pod = new AirtableExportPod();
        const srcHierarchy = "foo";
        timestamp = Time.now().toMillis() - 50000000;
        fs.writeFileSync(checkpoint, timestamp.toString(), {
          encoding: "utf8",
        });
        const stub = sinon
          .stub(pod, "processNote")
          .returns(Promise.resolve({ created: [], updated: [] }));
        const resp = await pod.execute({
          engine,
          vaults,
          wsRoot,
          config: {
            dest: "TODO",
            apiKey: "apikey",
            baseId: "baseId",
            tableName: "Dendron",
            srcFieldMapping: {
              Title: "title",
              "Updated On": "updated",
              Notes: "body",
            },
            srcHierarchy,
          },
        });
        expect(resp.notes).not.toBeNull();
        expect(stub.calledOnce).toBeTruthy();
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
      }
    );
  });
});
