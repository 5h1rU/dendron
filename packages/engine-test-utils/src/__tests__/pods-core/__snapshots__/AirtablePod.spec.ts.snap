// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airtable export pod basic basic 1`] = `
Array [
  Object {
    "anchors": Object {
      "welcome-to-dendron": Object {
        "column": 0,
        "line": 7,
        "text": "Welcome to Dendron",
        "type": "header",
        "value": "welcome-to-dendron",
      },
    },
    "body": "# Welcome to Dendron

This is the root of your dendron vault. If you decide to publish your entire vault, this will be your landing page. You are free to customize any part of this page except the frontmatter on top. 
",
    "children": Array [
      "bar",
      "foo",
    ],
    "contentHash": "ca76139e67ba2b9c5623ab2f05817238",
    "created": 1632958947884,
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "gQP8NR9nDgKGN6j0vFtND",
    "links": Array [],
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": 1632958947884,
    "vault": Object {
      "fsPath": "vault1",
    },
  },
  Object {
    "anchors": Object {},
    "body": "bar body
",
    "children": Array [],
    "contentHash": "6657f8cf533cd621d3f4219eb0f642b7",
    "created": 1,
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "bar",
    "id": "bar",
    "links": Array [],
    "parent": "gQP8NR9nDgKGN6j0vFtND",
    "title": "Bar",
    "type": "note",
    "updated": 1,
    "vault": Object {
      "fsPath": "vault1",
    },
  },
  Object {
    "anchors": Object {},
    "body": "foo body
",
    "children": Array [
      "foo.ch1",
    ],
    "contentHash": "664ac8773804ef0eac60e141f4da51cb",
    "created": 1,
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo",
    "id": "foo",
    "links": Array [],
    "parent": "gQP8NR9nDgKGN6j0vFtND",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "foo",
    },
    "title": "Foo",
    "type": "note",
    "updated": 1,
    "vault": Object {
      "fsPath": "vault1",
    },
  },
  Object {
    "anchors": Object {},
    "body": "foo.ch1 body
",
    "children": Array [],
    "contentHash": "0e984be1c58e3d9bd63d02cfdc83c978",
    "created": 1,
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.ch1",
    "id": "foo.ch1",
    "links": Array [],
    "parent": "foo",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "ch1",
    },
    "title": "Ch1",
    "type": "note",
    "updated": 1,
    "vault": Object {
      "fsPath": "vault1",
    },
  },
  Object {
    "anchors": Object {
      "welcome-to-dendron": Object {
        "column": 0,
        "line": 7,
        "text": "Welcome to Dendron",
        "type": "header",
        "value": "welcome-to-dendron",
      },
    },
    "body": "# Welcome to Dendron

This is the root of your dendron vault. If you decide to publish your entire vault, this will be your landing page. You are free to customize any part of this page except the frontmatter on top. 
",
    "children": Array [],
    "contentHash": "2a69e7fa06e1872f9f58a09c18df5577",
    "created": 1632958947884,
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "cAwfckLDTS2rcupArIRmi",
    "links": Array [],
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": 1632958947884,
    "vault": Object {
      "fsPath": "vault2",
    },
  },
  Object {
    "anchors": Object {
      "welcome-to-dendron": Object {
        "column": 0,
        "line": 7,
        "text": "Welcome to Dendron",
        "type": "header",
        "value": "welcome-to-dendron",
      },
    },
    "body": "# Welcome to Dendron

This is the root of your dendron vault. If you decide to publish your entire vault, this will be your landing page. You are free to customize any part of this page except the frontmatter on top. 
",
    "children": Array [],
    "contentHash": "9685fba144136005df054d36e714d7a4",
    "created": 1632958947884,
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "uplTjoqGorvARrJLiaepm",
    "links": Array [],
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": 1632958947884,
    "vault": Object {
      "fsPath": "vault3",
      "name": "vaultThree",
    },
  },
]
`;
